udev/zvol_id.c

```diff
@@ -50,80 +48,37 @@ const char *__asan_default_options(void) {
 }
 #endif
 
-static int
-ioctl_get_msg(char *var, int fd)
-{
-	int ret;
-	char msg[ZFS_MAX_DATASET_NAME_LEN];
-
-	ret = ioctl(fd, BLKZNAME, msg);
-	if (ret < 0) {
-		return (ret);
-	}
-
-	snprintf(var, ZFS_MAX_DATASET_NAME_LEN, "%s", msg);
-	return (ret);
-}
-
 int
-main(int argc, char **argv)
+main(int argc, const char *const *argv)
 {
-	int fd = -1, ret = 0, status = EXIT_FAILURE;
-	char zvol_name[ZFS_MAX_DATASET_NAME_LEN];
-	char *zvol_name_part = NULL;
-	char *dev_name;
-	struct stat64 statbuf;
-	int dev_minor, dev_part;
-	int i;
-
-	if (argc < 2) {
-		fprintf(stderr, "Usage: %s /dev/zvol_device_node\n", argv[0]);
-		goto fail;
+	if (argc != 2) {
+		fprintf(stderr, "usage: %s /dev/zdX\n", argv[0]);
+		return (1);
 	}
-
-	dev_name = argv[1];
-	ret = stat64(dev_name, &statbuf);
-	if (ret != 0) {
-		fprintf(stderr, "Unable to access device file: %s\n", dev_name);
-		goto fail;
+	const char *dev_name = argv[1];
+
+	int fd;
+	struct stat sb;
+	if ((fd = open(dev_name, O_RDONLY|O_CLOEXEC)) == -1 ||
+	    fstat(fd, &sb) != 0) {
+		fprintf(stderr, "%s: %s\n", dev_name, strerror(errno));
+		return (1);
 	}
 
-	dev_minor = minor(statbuf.st_rdev);
-	dev_part = dev_minor % ZVOL_MINORS;
-
-	fd = open(dev_name, O_RDONLY);
-	if (fd < 0) {
-		fprintf(stderr, "Unable to open device file: %s\n", dev_name);
-		goto fail;
-	}
-
-	ret = ioctl_get_msg(zvol_name, fd);
-	if (ret < 0) {
-		fprintf(stderr, "ioctl_get_msg failed: %s\n", strerror(errno));
-		goto fail;
-	}
-	if (dev_part > 0)
-		ret = asprintf(&zvol_name_part, "%s-part%d", zvol_name,
-		    dev_part);
-	else
-		ret = asprintf(&zvol_name_part, "%s", zvol_name);
-
-	if (ret == -1 || zvol_name_part == NULL)
-		goto fail;
-
-	for (i = 0; i < strlen(zvol_name_part); i++) {
-		if (isblank(zvol_name_part[i]))
-			zvol_name_part[i] = '+';
+	char zvol_name[MAXNAMELEN + strlen("-part") + 10];
+	if (ioctl(fd, BLKZNAME, zvol_name) == -1) {
+		fprintf(stderr, "%s: BLKZNAME: %s\n",
+		    dev_name, strerror(errno));
+		return (1);
 	}
 
-	printf("%s\n", zvol_name_part);
-	status = EXIT_SUCCESS;
+	unsigned int dev_part = minor(sb.st_rdev) % ZVOL_MINORS;
+	if (dev_part != 0)
+		sprintf(zvol_name + strlen(zvol_name), "-part%u", dev_part);
 
-fail:
-	if (zvol_name_part)
-		free(zvol_name_part);
-	if (fd >= 0)
-		close(fd);
+	for (size_t i = 0; i < strlen(zvol_name); ++i)
+		if (isblank(zvol_name[i]))
+			zvol_name[i] = '+';
 
-	return (status);
+	puts(zvol_name);
```

Comment:
Even though it's optional in C99 let's add the missing `return (0);` here.

Reply:
Fixed